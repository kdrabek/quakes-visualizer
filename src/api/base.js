import 'whatwg-fetch';
import queryString from 'query-string';
import Immutable from 'immutable';


function parse(resp) {
  return new Promise((resolve, reject) => {
    resp.json().then(
      data => resolve({ data, status: resp.status }),
      error => reject(error),
    );
  });
}


function toImmutable(resp) {
  return Immutable.fromJS(resp.data.features);
}


function checkStatus(resp) {
  if (resp.status >= 400) {
    throw new Error(`${resp.status}: ${JSON.stringify(resp.data)}`);
  }
  return resp;
}


function processResponse(resp) {
  return resp.then(parse).then(checkStatus).then(toImmutable);
}


export function makeRequest(url, queryParams = {}, options = {}) {
  const { method = 'GET', data, headers = {} } = options;
  const requestHeaders = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...headers,
  };

  const params = `${queryParams ? '?' : ''}${queryString.stringify(queryParams)}`;
  const mergedUrl = `${url}${params}`;

  let init = { method, header: requestHeaders };
  if (data) {
    init = { ...init, body: JSON.stringify(data) };
  }

  return processResponse(fetch(mergedUrl, init));
}

export default {
  makeRequest,
};
