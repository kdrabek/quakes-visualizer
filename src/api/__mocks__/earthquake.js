import Immutable from 'immutable';

export const mockedResponse = {};
const get = jest.fn(async () => Immutable.fromJS(mockedResponse));

export default {
  get,
};
