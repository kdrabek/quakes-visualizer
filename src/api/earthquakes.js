import { earthquakeApiUrl } from '../configs';
import { makeRequest } from './base';


const get = (startDate, endDate, minmagnitude) =>
  makeRequest(
    `${earthquakeApiUrl}/query`,
    {
      format: 'geojson',
      starttime: startDate,
      endtime: endDate,
      minmagnitude,
    },
  );

export default {
  get,
};
