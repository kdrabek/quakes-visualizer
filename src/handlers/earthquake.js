import Immutable from 'immutable';

import { createAction, handleActions } from 'redux-actions';
import api from '../api';

const getEarthquakes = createAction(
  'GET_EARTHQUAKES',
  (startDate, endDate, minmagnitude) =>
    api.getEarthquake(startDate, endDate, minmagnitude),
);

const actions = { getEarthquakes };

const initialState = Immutable.fromJS({
  data: [],
  pending: false,
  error: false,
});

const reducer = handleActions({
  GET_EARTHQUAKES_PENDING: state => state
    .update('pending', () => true),

  GET_EARTHQUAKES_FULFILLED: (state, action) => state
    .update('pending', () => false)
    .set('data', Immutable.fromJS(action.payload)),

  GET_EARTHQUAKES_REJECTED: (state, action) => state
    .update('pending', () => false)
    .set('error', Immutable.fromJS(action.payload)),

}, initialState);


export default {
  actions,
  reducer,
};
