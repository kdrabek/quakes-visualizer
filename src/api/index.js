import earthquakes from './earthquakes';

export default {
  getEarthquake: earthquakes.get,
};
