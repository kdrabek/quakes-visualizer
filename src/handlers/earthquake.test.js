/* eslint-disable import/first */
jest.mock('../api/earthquakes');
import api from '../api/earthquakes';
import Immutable from 'immutable';
import { combineReducers } from 'redux-immutable';

import earthquake from './earthquake';


async function fulfilledAction(action) {
  return {
    ...action,
    payload: await action.payload,
    type: `${action.type}_FULFILLED`,
  };
}

async function pendingAction(action) {
  return {
    ...action,
    payload: {},
    type: `${action.type}_PENDING`,
  };
}

async function rejectedAction(action) {
  return {
    ...action,
    payload: { error: 'Something went wrong' },
    type: `${action.type}_REJECTED`,
  };
}

describe('handlers', () => {
  const reducer = combineReducers({ earthquakes: earthquake.reducer });
  const state = reducer(undefined, Immutable.Map());

  it('has proper default state', () => {
    expect(state.toJS()).toMatchSnapshot();
  });

  describe('GET_EARTHQUAKES action getting new data', () => {
    const resp = earthquake.actions.getEarthquakes(new Date('2018-03-20'), 5);

    it('api.getEarthquakes called once', async () => {
      expect(api.get.mock.calls.length).toBe(1);
    });

    it('has right data in the payload', async () => {
      expect(await resp.payload).toMatchSnapshot();
    });

    it('has data in right place when action FULFILLED', async () => {
      const newState = reducer(state, await fulfilledAction(resp));

      expect(newState).toMatchSnapshot();
    });

    it('has data in right place when action PENDING', async () => {
      const newState = reducer(state, await pendingAction(resp));

      expect(newState).toMatchSnapshot();
    });

    it('has data in right place when action REJECTED', async () => {
      const newState = reducer(state, await rejectedAction(resp));

      expect(newState).toMatchSnapshot();
    });
  });

  describe('UPDATE_MAP action getting new data', () => {
    const stateWithCoords = reducer(Immutable.fromJS({
      earthquakes: {
        data: [
          {
            geometry: { coordinates: [167.249, -13.8633] },
            id: 'usc000lvb5'
          },
        ],
      },
    }), Immutable.Map());

    const resp = earthquake.actions.updateMap('usc000lvb5');

    it('has data in right place', async () => {
      const newState = reducer(stateWithCoords, resp);

      expect(newState).toMatchSnapshot();
    });
  });
});
