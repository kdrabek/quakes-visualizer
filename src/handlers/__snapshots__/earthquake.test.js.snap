// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers GET_EARTHQUAKES action getting new data has data in right place when action FULFILLED 1`] = `
Immutable.Map {
  "earthquakes": Immutable.Map {
    "data": Immutable.List [
      Immutable.Map {
        "type": "Feature",
        "properties": Immutable.Map {
          "tsunami": 1,
          "tz": 660,
          "place": "32km W of Sola, Vanuatu",
          "felt": null,
          "cdi": null,
          "mmi": 4.84,
          "code": "c000lvb5",
          "time": 1388592209000,
          "magType": "mww",
          "dmin": 3.997,
          "url": "https://earthquake.usgs.gov/earthquakes/eventpage/usc000lvb5",
          "nst": null,
          "detail": "https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=usc000lvb5&format=geojson",
          "sig": 650,
          "status": "reviewed",
          "mag": 6.5,
          "gap": 14,
          "rms": 0.76,
          "title": "M 6.5 - 32km W of Sola, Vanuatu",
          "ids": ",at00myqcls,pt14001000,usc000lvb5,",
          "sources": ",at,pt,us,",
          "types": ",cap,geoserve,impact-link,losspager,moment-tensor,moment-tensor,moment-tensor,moment-tensor,nearby-cities,origin,phase-data,shakemap,tectonic-summary,",
          "type": "earthquake",
          "updated": 1394151955000,
          "alert": "green",
          "net": "us",
        },
        "geometry": Immutable.Map {
          "type": "Point",
          "coordinates": Immutable.List [
            167.249,
            -13.8633,
            187,
          ],
        },
        "id": "usc000lvb5",
      },
    ],
    "pending": false,
    "error": false,
    "mapCoords": Immutable.Map {
      "lng": 167.249,
      "lat": -13.8633,
    },
  },
}
`;

exports[`handlers GET_EARTHQUAKES action getting new data has data in right place when action PENDING 1`] = `
Immutable.Map {
  "earthquakes": Immutable.Map {
    "data": Immutable.List [],
    "pending": true,
    "error": false,
    "mapCoords": null,
  },
}
`;

exports[`handlers GET_EARTHQUAKES action getting new data has data in right place when action REJECTED 1`] = `
Immutable.Map {
  "earthquakes": Immutable.Map {
    "data": Immutable.List [],
    "pending": false,
    "error": Immutable.Map {
      "error": "Something went wrong",
    },
    "mapCoords": null,
  },
}
`;

exports[`handlers GET_EARTHQUAKES action getting new data has right data in the payload 1`] = `
Immutable.List [
  Immutable.Map {
    "type": "Feature",
    "properties": Immutable.Map {
      "tsunami": 1,
      "tz": 660,
      "place": "32km W of Sola, Vanuatu",
      "felt": null,
      "cdi": null,
      "mmi": 4.84,
      "code": "c000lvb5",
      "time": 1388592209000,
      "magType": "mww",
      "dmin": 3.997,
      "url": "https://earthquake.usgs.gov/earthquakes/eventpage/usc000lvb5",
      "nst": null,
      "detail": "https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=usc000lvb5&format=geojson",
      "sig": 650,
      "status": "reviewed",
      "mag": 6.5,
      "gap": 14,
      "rms": 0.76,
      "title": "M 6.5 - 32km W of Sola, Vanuatu",
      "ids": ",at00myqcls,pt14001000,usc000lvb5,",
      "sources": ",at,pt,us,",
      "types": ",cap,geoserve,impact-link,losspager,moment-tensor,moment-tensor,moment-tensor,moment-tensor,nearby-cities,origin,phase-data,shakemap,tectonic-summary,",
      "type": "earthquake",
      "updated": 1394151955000,
      "alert": "green",
      "net": "us",
    },
    "geometry": Immutable.Map {
      "type": "Point",
      "coordinates": Immutable.List [
        167.249,
        -13.8633,
        187,
      ],
    },
    "id": "usc000lvb5",
  },
]
`;

exports[`handlers UPDATE_MAP action getting new data has data in right place 1`] = `
Immutable.Map {
  "earthquakes": Immutable.Map {
    "data": Immutable.List [
      Immutable.Map {
        "geometry": Immutable.Map {
          "coordinates": Immutable.List [
            167.249,
            -13.8633,
          ],
        },
        "id": "usc000lvb5",
      },
    ],
    "mapCoords": Immutable.Map {
      "lng": 167.249,
      "lat": -13.8633,
    },
  },
}
`;

exports[`handlers has proper default state 1`] = `
Object {
  "earthquakes": Object {
    "data": Array [],
    "error": false,
    "mapCoords": null,
    "pending": false,
  },
}
`;
